File: go/lessons.go
package main

import (
	"database/sql"
	"fmt"
	"strings"
	"time"
)

type Lesson struct {
	ID          int64
	Category    string
	Title       string
	Content     string
	ProjectPath string
	Tags        string
	Source      string
	SourceRef   string
	RecallCount int
	CreatedAt   float64
	UpdatedAt   float64
}

func scanLessons(rows *sql.Rows) ([]Lesson, error) {
	var lessons []Lesson
	for rows.Next() {
		var l Lesson
		if err := rows.Scan(&l.ID, &l.Category, &l.Title, &l.Content,
			&l.ProjectPath, &l.Tags, &l.Source, &l.SourceRef,
			&l.RecallCount, &l.CreatedAt, &l.UpdatedAt); err != nil {
			return nil, err
		}
		lessons = append(lessons, l)
	}
	return lessons, rows.Err()
}

func normalizeTags(tags string) string {
	parts := strings.Split(tags, ",")
	var normalized []string
	for _, p := range parts {
		t := strings.TrimSpace(strings.ToLower(p))
		if t != "" {
			normalized = append(normalized, t)
		}
	}
	return strings.Join(normalized, ",")
}

func saveLesson(db *sql.DB, l Lesson) (int64, error) {
	now := float64(time.Now().Unix())
	l.Tags = normalizeTags(l.Tags)
	res, err := db.Exec(
		`INSERT INTO lessons (category, title, content, project_path, tags, source, source_ref, recall_count, created_at, updated_at)
		 VALUES (?, ?, ?, ?, ?, ?, ?, 0, ?, ?)`,
		l.Category, l.Title, l.Content, l.ProjectPath, l.Tags, l.Source, l.SourceRef, now, now,
	)
	if err != nil {
		return 0, err
	}
	return res.LastInsertId()
}

func updateLesson(db *sql.DB, id int64, title, content, category, tags *string) error {
	now := float64(time.Now().Unix())

	setClauses := []string{"updated_at = ?"}
	args := []any{now}

	if title != nil {
		setClauses = append(setClauses, "title = ?")
		args = append(args, *title)
	}
	if content != nil {
		setClauses = append(setClauses, "content = ?")
		args = append(args, *content)
	}
	if category != nil {
		setClauses = append(setClauses, "category = ?")
		args = append(args, *category)
	}
	if tags != nil {
		normalized := normalizeTags(*tags)
		setClauses = append(setClauses, "tags = ?")
		args = append(args, normalized)
	}

	args = append(args, id)
	query := fmt.Sprintf("UPDATE lessons SET %s WHERE id = ?", strings.Join(setClauses, ", "))
	res, err := db.Exec(query, args...)
	if err != nil {
		return err
	}
	n, _ := res.RowsAffected()
	if n == 0 {
		return fmt.Errorf("lesson not found: %d", id)
	}
	return nil
}

func getLessonByID(db *sql.DB, id int64) (*Lesson, error) {
	row := db.QueryRow(
		`SELECT id, category, title, content, project_path, tags, source, source_ref, recall_count, created_at, updated_at
		 FROM lessons WHERE id = ?`, id)
	var l Lesson
	err := row.Scan(&l.ID, &l.Category, &l.Title, &l.Content,
		&l.ProjectPath, &l.Tags, &l.Source, &l.SourceRef,
		&l.RecallCount, &l.CreatedAt, &l.UpdatedAt)
	if err != nil {
		return nil, err
	}
	return &l, nil
}

func deleteLesson(db *sql.DB, id int64) error {
	res, err := db.Exec("DELETE FROM lessons WHERE id = ?", id)
	if err != nil {
		return err
	}
	n, _ := res.RowsAffected()
	if n == 0 {
		return fmt.Errorf("lesson not found: %d", id)
	}
	return nil
}

func listLessons(db *sql.DB, projectPath, category string, limit int) ([]Lesson, error) {
	query := `SELECT id, category, title, content, project_path, tags, source, source_ref, recall_count, created_at, updated_at
		 FROM lessons WHERE 1=1`
	var args []any

	if projectPath != "" {
		query += " AND (project_path = ? OR project_path = '')"
		args = append(args, projectPath)
	}
	if category != "" {
		query += " AND category = ?"
		args = append(args, category)
	}

	query += " ORDER BY recall_count DESC, created_at DESC LIMIT ?"
	args = append(args, limit)

	rows, err := db.Query(query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanLessons(rows)
}

func searchLessons(db *sql.DB, query, projectPath, category string, limit int) ([]Lesson, error) {
	var results []Lesson
	var err error

	if len(query) >= 3 {
		results, err = fts5SearchLessons(db, query, limit*2)
		if err != nil || len(results) == 0 {
			results, err = likeSearchLessons(db, query, limit*2)
		}
	} else {
		results, err = likeSearchLessons(db, query, limit*2)
	}
	if err != nil {
		return nil, err
	}

	// Apply filters
	filtered := results[:0]
	for _, l := range results {
		if projectPath != "" && l.ProjectPath != "" && l.ProjectPath != projectPath {
			continue
		}
		if category != "" && l.Category != category {
			continue
		}
		filtered = append(filtered, l)
	}

	if len(filtered) > limit {
		filtered = filtered[:limit]
	}
	return filtered, nil
}

func fts5SearchLessons(db *sql.DB, query string, limit int) ([]Lesson, error) {
	rows, err := db.Query(
		`SELECT l.id, l.category, l.title, l.content, l.project_path, l.tags, l.source, l.source_ref, l.recall_count, l.created_at, l.updated_at
		 FROM lessons_fts f JOIN lessons l ON l.id = f.rowid
		 WHERE lessons_fts MATCH ? ORDER BY rank LIMIT ?`, query, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanLessons(rows)
}

func likeSearchLessons(db *sql.DB, query string, limit int) ([]Lesson, error) {
	pattern := "%" + query + "%"
	rows, err := db.Query(
		`SELECT id, category, title, content, project_path, tags, source, source_ref, recall_count, created_at, updated_at
		 FROM lessons WHERE title LIKE ? OR content LIKE ? OR tags LIKE ?
		 ORDER BY recall_count DESC, created_at DESC LIMIT ?`,
		pattern, pattern, pattern, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanLessons(rows)
}

func incrementRecallCount(db *sql.DB, ids []int64) {
	if len(ids) == 0 {
		return
	}
	placeholders := make([]string, len(ids))
	args := make([]any, len(ids))
	for i, id := range ids {
		placeholders[i] = "?"
		args[i] = id
	}
	query := fmt.Sprintf("UPDATE lessons SET recall_count = recall_count + 1 WHERE id IN (%s)",
		strings.Join(placeholders, ","))
	db.Exec(query, args...)
}

func categoryLabel(category string) string {
	switch category {
	case "review_feedback":
		return "レビュー指摘"
	case "design_decision":
		return "設計決定"
	case "lesson_learned":
		return "学び"
	case "pattern":
		return "パターン"
	case "mistake":
		return "失敗"
	default:
		return category
	}
}

func searchLessonsByTag(db *sql.DB, tag, projectPath string, limit int) ([]Lesson, error) {
	rows, err := db.Query(
		`SELECT id, category, title, content, project_path, tags, source, source_ref, recall_count, created_at, updated_at
		 FROM lessons
		 WHERE tags LIKE ?
		 AND (project_path = ? OR project_path = '')
		 ORDER BY recall_count DESC, created_at DESC LIMIT ?`,
		"%"+tag+"%", projectPath, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanLessons(rows)
}

// extractSearchSegments splits text into overlapping rune-based segments
// for trigram-friendly FTS5 search. Full strings don't match paraphrased text,
// but short segments (e.g. "設計に固") can match across similar titles.
func extractSearchSegments(text string, segLen int) []string {
	runes := []rune(text)
	if len(runes) <= segLen {
		if len(runes) >= 3 {
			return []string{text}
		}
		return nil
	}
	var segments []string
	step := segLen
	for i := 0; i+segLen <= len(runes); i += step {
		segments = append(segments, string(runes[i:i+segLen]))
	}
	return segments
}

func findSimilarLessons(db *sql.DB, title, content, projectPath string, excludeID int64, limit int) ([]Lesson, error) {
	var results []Lesson
	seen := map[int64]bool{excludeID: true}

	// Search with short segments of the title (5 runes each)
	segments := extractSearchSegments(title, 5)
	for _, seg := range segments {
		candidates, err := fts5SearchLessons(db, seg, limit*3)
		if err == nil {
			for _, l := range candidates {
				if !seen[l.ID] {
					seen[l.ID] = true
					results = append(results, l)
				}
			}
		}
		if len(results) >= limit {
			break
		}
	}

	// Fallback: search with content segments if not enough results
	if len(results) < limit {
		contentSegments := extractSearchSegments(content, 5)
		for _, seg := range contentSegments {
			candidates, err := fts5SearchLessons(db, seg, limit*3)
			if err == nil {
				for _, l := range candidates {
					if !seen[l.ID] {
						seen[l.ID] = true
						results = append(results, l)
					}
				}
			}
			if len(results) >= limit {
				break
			}
		}
	}

	filtered := results[:0]
	for _, l := range results {
		if projectPath == "" || l.ProjectPath == "" || l.ProjectPath == projectPath {
			filtered = append(filtered, l)
		}
	}

	if len(filtered) > limit {
		filtered = filtered[:limit]
	}
	return filtered, nil
}

File: ts/src/lessons.ts
import type { DB } from "./db.js";

export interface Lesson {
  id: number;
  category: string;
  title: string;
  content: string;
  project_path: string;
  tags: string;
  source: string;
  source_ref: string;
  recall_count: number;
  created_at: number;
  updated_at: number;
}

export interface LessonInput {
  category: string;
  title: string;
  content: string;
  project_path: string;
  tags: string;
  source: string;
  source_ref: string;
}

export interface LessonUpdate {
  title?: string;
  content?: string;
  category?: string;
  tags?: string;
}

export function normalizeTags(tags: string): string {
  return tags
    .split(",")
    .map((t) => t.trim().toLowerCase())
    .filter((t) => t !== "")
    .join(",");
}

export function saveLesson(db: DB, l: LessonInput): number {
  const now = Math.floor(Date.now() / 1000);
  const tags = normalizeTags(l.tags);
  const result = db
    .prepare(
      `INSERT INTO lessons (category, title, content, project_path, tags, source, source_ref, recall_count, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, 0, ?, ?)`,
    )
    .run(
      l.category,
      l.title,
      l.content,
      l.project_path,
      tags,
      l.source,
      l.source_ref,
      now,
      now,
    );
  return Number(result.lastInsertRowid);
}

export function updateLesson(db: DB, id: number, patch: LessonUpdate): void {
  const now = Math.floor(Date.now() / 1000);
  const setClauses: string[] = ["updated_at = ?"];
  const args: (string | number)[] = [now];

  if (patch.title !== undefined) {
    setClauses.push("title = ?");
    args.push(patch.title);
  }
  if (patch.content !== undefined) {
    setClauses.push("content = ?");
    args.push(patch.content);
  }
  if (patch.category !== undefined) {
    setClauses.push("category = ?");
    args.push(patch.category);
  }
  if (patch.tags !== undefined) {
    setClauses.push("tags = ?");
    args.push(normalizeTags(patch.tags));
  }
  args.push(id);

  const result = db
    .prepare(`UPDATE lessons SET ${setClauses.join(", ")} WHERE id = ?`)
    .run(...args);
  if (result.changes === 0) {
    throw new Error(`lesson not found: ${id}`);
  }
}

export function deleteLesson(db: DB, id: number): void {
  const result = db.prepare("DELETE FROM lessons WHERE id = ?").run(id);
  if (result.changes === 0) {
    throw new Error(`lesson not found: ${id}`);
  }
}

export function listLessons(
  db: DB,
  projectPath: string,
  category: string,
  limit: number,
): Lesson[] {
  let q = `SELECT id, category, title, content, project_path, tags, source, source_ref, recall_count, created_at, updated_at
       FROM lessons WHERE 1=1`;
  const args: (string | number)[] = [];
  if (projectPath !== "") {
    q += " AND (project_path = ? OR project_path = '')";
    args.push(projectPath);
  }
  if (category !== "") {
    q += " AND category = ?";
    args.push(category);
  }
  q += " LIMIT ?";
  args.push(limit);
  return db.prepare(q).all(...args) as Lesson[];
}

function fts5SearchLessons(db: DB, query: string, limit: number): Lesson[] {
  return db
    .prepare(
      `SELECT l.id, l.category, l.title, l.content, l.project_path, l.tags, l.source, l.source_ref, l.recall_count, l.created_at, l.updated_at
       FROM lessons_fts f JOIN lessons l ON l.id = f.rowid
       WHERE lessons_fts MATCH ? ORDER BY rank LIMIT ?`,
    )
    .all(query, limit) as Lesson[];
}

function likeSearchLessons(db: DB, query: string, limit: number): Lesson[] {
  const pattern = `%${query}%`;
  return db
    .prepare(
      `SELECT id, category, title, content, project_path, tags, source, source_ref, recall_count, created_at, updated_at
       FROM lessons WHERE title LIKE ? OR content LIKE ? OR tags LIKE ?
       LIMIT ?`,
    )
    .all(pattern, pattern, pattern, limit) as Lesson[];
}

export function searchLessons(
  db: DB,
  query: string,
  projectPath: string,
  category: string,
  tags: string,
  source: string,
  limit: number,
): Lesson[] {
  let results: Lesson[];
  if (query.length >= 3) {
    try {
      results = fts5SearchLessons(db, query, limit * 2);
      if (results.length === 0) {
        results = likeSearchLessons(db, query, limit * 2);
      }
    } catch {
      results = likeSearchLessons(db, query, limit * 2);
    }
  } else {
    results = likeSearchLessons(db, query, limit * 2);
  }

  const tagFilters = tags
    ? tags
        .split(",")
        .map((t) => t.trim().toLowerCase())
        .filter((t) => t !== "")
    : [];

  const filtered = results.filter((l) => {
    if (projectPath !== "" && l.project_path !== "" && l.project_path !== projectPath)
      return false;
    if (category !== "" && l.category !== category) return false;
    if (source !== "" && l.source !== source) return false;
    if (tagFilters.length > 0) {
      const lower = l.tags.toLowerCase();
      if (!tagFilters.some((tf) => lower.includes(tf))) return false;
    }
    return true;
  });

  return filtered.slice(0, limit);
}

export function incrementRecallCount(db: DB, ids: number[]): void {
  if (ids.length === 0) return;
  const placeholders = ids.map(() => "?").join(",");
  db.prepare(
    `UPDATE lessons SET recall_count = recall_count + 1 WHERE id IN (${placeholders})`,
  ).run(...ids);
}

export function categoryLabel(category: string): string {
  switch (category) {
    case "review_feedback":
      return "レビュー指摘";
    case "design_decision":
      return "設計決定";
    case "lesson_learned":
      return "学び";
    case "pattern":
      return "パターン";
    case "mistake":
      return "失敗";
    default:
      return category;
  }
}

function extractSearchSegments(text: string, segLen: number): string[] {
  const runes = Array.from(text);
  if (runes.length <= segLen) {
    return runes.length >= 3 ? [text] : [];
  }
  const segments: string[] = [];
  for (let i = 0; i + segLen <= runes.length; i += segLen) {
    segments.push(runes.slice(i, i + segLen).join(""));
  }
  return segments;
}

export function findSimilarLessons(
  db: DB,
  title: string,
  content: string,
  projectPath: string,
  excludeID: number,
  limit: number,
): Lesson[] {
  const seen = new Set<number>([excludeID]);
  const results: Lesson[] = [];

  for (const seg of extractSearchSegments(title, 5)) {
    try {
      const candidates = fts5SearchLessons(db, seg, limit * 3);
      for (const l of candidates) {
        if (!seen.has(l.id)) {
          seen.add(l.id);
          results.push(l);
        }
      }
    } catch {
      // ignore segment errors
    }
    if (results.length >= limit) break;
  }

  if (results.length < limit) {
    for (const seg of extractSearchSegments(content, 5)) {
      try {
        const candidates = fts5SearchLessons(db, seg, limit * 3);
        for (const l of candidates) {
          if (!seen.has(l.id)) {
            seen.add(l.id);
            results.push(l);
          }
        }
      } catch {
        // ignore
      }
      if (results.length >= limit) break;
    }
  }

  const filtered = results.filter(
    (l) => projectPath === "" || l.project_path === "" || l.project_path === projectPath,
  );
  return filtered.slice(0, limit);
}
